test/extension.test.ts

```diff
@@ -244,53 +257,53 @@ suite("Extension Tests", function() {
 
         await checkCoverage(() => {
             // Look for exact coverage on the file
-            const javaCachedLines: ICoverageLines = getCachedLines();
+            const javaCachedLines: ICoverageLines = decorationSpy.getCall(1).args[1];
             expect(javaCachedLines.full.length).to.equal(4);
             expect(javaCachedLines.none.length).to.equal(3);
         });
 
-        extension.exports.emptyLastCoverage();
+        decorationSpy.restore();
         return vscode.commands.executeCommand("coverage-gutters.removeWatch");
     });
 
     test(
         "Run coverage and open files to see line coverage percentage in the status bar @integration",
         async () => {
-        await vscode.commands.executeCommand("workbench.action.closeAllEditors");
-        const setCoverageSpy = sinon.spy(StatusBarToggler.prototype, "setCoverage");
+            await vscode.commands.executeCommand("workbench.action.closeAllEditors");
+            const setCoverageSpy = sinon.spy(StatusBarToggler.prototype, "setCoverage");
```

Comment:
@mattseddon Followed along with what you had here but sadly I wasn't able to spy on private typescript functions so I had to change `setDecorationsForEditor` to public 🤔 .

Reply:
I'll check this out and take a look tonight. 

One quick things that I could suggest is that you can call `restore` in an `afterEach` hook instead of restoring each stub manually.

Reply:
Actually, you might not be able to do that because of your global stub 🤦🏻‍♂️

Reply:
@mattseddon Yah I originally had that plan to do the restore / setup in a sandbox inside the before / after.
Wasn't able to figure out how to appease typescript though and it didn't know that the sinon spy was a union type between a `spy | config`.

I ended up just putting them inline as a first pass to get everything working 🤔 .

Reply:
After I switched the stub config to an instance (using `const stubConfig = sinon.createStubInstance(Config) as Config;`) I was able to remove the inline calls to restore and add

```
	this.afterEach(() => restore());
```

at the top of this file and all of the tests still passed.

Reply:
> @mattseddon Followed along with what you had here but sadly I wasn't able to spy on private typescript functions so I had to change `setDecorationsForEditor` to public 🤔 .

I've had a look into this and you could probably get the visible editors from `window.visibleTextEditors` and then add spies to their `setDecorations` methods. That would give you a true integration test and make it possible to leave the method private. Seeing as there should only be one editor open you could probably even use `window.activeTextEditor` which would be easier.

If you go down that route I would definitely add an `afterEach` hook that runs:

```
commands.executeCommand('workbench.action.closeAllEditors')
```

again this should keep your tests independent. Up to you whether the effort is worth the benefit.

Reply:
RE Per test file stubConfig:
That make sense to me, I will try to refine this a bit more, good suggestions!

RE `Public` vs `Private`:
Excellent, I like the idea of using the vscode api for the spies as those have public functions that we use and as you mentioned create better integration tests anyways.

Reply:
@mattseddon https://github.com/ryanluker/vscode-coverage-gutters/pull/328/commits/e60053cbe07ae8650dbe53532713ec13d0d11b35
RE Config stub:
Ended up cleaning up the config stub a bit more but didn't go the whole way to having it per test case basis.

RE The spy stuff:
Sadly `setDecorations` was not very spyable via sinon on the different vscode provided areas.
I tried a bunch of different options using active and visible editors but nothing would let me spy the decorations setting function, ended up leaving as is for now.